src/coreclr/jit/objectalloc.cpp

```diff
@@ -717,7 +798,54 @@ void ObjectAllocator::UpdateAncestorTypes(GenTree* tree, ArrayStack<GenTree*>* p
                 keepChecking = true;
                 break;
 
+            case GT_COLON:
+            {
+                GenTree* const lhs = parent->AsOp()->gtGetOp1();
+                GenTree* const rhs = parent->AsOp()->gtGetOp2();
+
+                // One or both children may have been retyped.
+                // Ensure we don't lose the fact that the joint result
+                // may be a GC type.
+                //
+                var_types lhsType = lhs->TypeGet();
+                var_types rhsType = rhs->TypeGet();
+
+                if (lhsType == TYP_REF)
+                {
+                    assert(rhsType != TYP_REF);
+                    assert(rhsType == tree->TypeGet());
+                    newType = TYP_BYREF;
+                }
+
+                if (rhsType == TYP_REF)
+                {
+                    assert(lhsType != TYP_REF);
+                    assert(lhsType == tree->TypeGet());
+                    newType = TYP_BYREF;
+                }
+
+                tree->ChangeType(newType);
+
+                if (lhsType != newType)
+                {
+                    lhs->ChangeType(newType);
+                }
+
+                if (rhsType != newType)
+                {
+                    rhs->ChangeType(newType);
+                }
+
+                parent->ChangeType(newType);
```

Comment:
It still seems odd to me to forcefully retype an arbitrary sibling node of an ancestor, but maybe all `TYP_REF` JIT IR nodes really do have equivalently typed `TYP_BYREF` nodes.

Reply:
For example. this will change `TYP_REF LCL_VAR(local whose LclVarDsc is TYP_REF)` to `TYP_BYREF LCL_VAR(local whose LclVarDsc is TYP_REF)`... That will hit this assert:

https://github.com/dotnet/runtime/blob/fcc2f58b4185862f86c569a807d97194618fb5a0/src/coreclr/jit/lclvars.cpp#L4691-L4694

Not sure if it will actually cause issues downstream, but it feels like we should insert some form of cast to maintain IR invariants. Potentially in quite a few places in object stack allocation. Anyway, something to think about, we don't need to block on this given that `QMARK`s have very restricted shapes.

Reply:
Happy to reconsider ... I added this because of assertion failures but don't recall the specific case and now can't reproduce it. Not clear to me how the JIT could create a TYP_REF GT_COLON, at least not one that this code would see.

Reply:
Here's the motivating example -- from the object stack allocation tests 
```
               [000035] DA-X-------                         *  STORE_LCL_VAR ref    V03 tmp3
               [000034] ---X-------                         \--*  QMARK     ref
               [000025] -----------    if                      +--*  EQ        int
               [000024] -----------                            |  +--*  LCL_VAR   ref    V02 tmp2
               [000023] -----------                            |  \--*  CNS_INT   ref    null
               [000033] ---X-------    if                      \--*  COLON     ref
               [000031] ---X------- else                          +--*  QMARK     ref
               [000027] ---X-------    if                         |  +--*  NE        int
               [000026] #--X-------                               |  |  +--*  IND       long
               [000022] -----------                               |  |  |  \--*  LCL_VAR   ref    V02 tmp2
               [000021] H----------                               |  |  \--*  CNS_INT(h) long   0x7ffe34d5e970 class ObjectStackAllocation.SimpleClassB
               [000030] -----------    if                         |  \--*  COLON     ref
               [000028] ----------- else                          |     +--*  LCL_VAR   ref    V02 tmp2
               [000029] ----------- then                          |     \--*  CNS_INT   ref    null
               [000032] ----------- then                          \--*  CNS_INT   ref    null
```
We stack allocate the ref class, and so retype `tmp2` to byref. This propagates up and if we don't retype sibling `[000029]` we hit asserts.

I think it may be the case that the mistyped sibling must be null... I can check for that.